### PR TITLE
feat(github): set package registry credentials automatically

### LINKED ISSUE #17705: Support automatic GitHub Packages permissions with platform token
### What would you like Renovate to be able to do?

The Renovate App currently doesn't ask for Packages permissions:
image

Before we ask for that permission (which will trigger a lot of permission notification emails) I want to make sure it *works*, ideally for both github.com and self-hosted GHES at the same time.

For github.com, it seems like *.pkg.github.com and ghcr.io are the relevant hosts which users would like to work. Do github tokens typically work for both? e.g. can we assume that if we have a github.com token it also might be valid for ghcr.io too?

### If you have any ideas on how this should be implemented, please tell us here.

Might need addition of new hostRules as part of initPlatform()

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Add credentials for *.pkg.github.com automatically

## Context

Closes #17705

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/github/index.ts [hunk 1]
```diff
@@ -214,6 +214,25 @@ export async function initPlatform({
         password: token.replace(/^x-access-token:/, ''),
       },
     ];
+    logger.debug('Adding GitHub token as npm.pkg.github.com Basic token');
+    platformResult.hostRules.push({
+      matchHost: 'npm.pkg.github.com',
+      hostType: 'npm',
+      token: token.replace(/^x-access-token:/, ''),
+      authType: 'Basic',
```
Comment (viceice) [inline lib/modules/platform/github/index.ts:222]: are you sure about that? 🤔

**Review by viceice — COMMENTED** [review]
failed tests

File: lib/modules/platform/github/index.ts (hunk 1 above)
Comment (rarkins) [reply lib/modules/platform/github/index.ts:222]: Check the associated issue and let me know what you think

File: lib/modules/platform/github/index.md [hunk 2]
```diff
@@ -85,6 +85,16 @@ The [GitHub App associated email](https://github.community/t/logging-into-git-as
 It needs to have the user id _and_ the username followed by the `users.noreply.`-domain of either github.com or the GitHub Enterprise Server.
 A way to get the user id of a GitHub app is to [query the user API](https://docs.github.com/en/rest/reference/users#get-a-user) at `api.github.com/users/self-hosted-renovate[bot]` (github.com) or `github.enterprise.com/api/v3/users/self-hosted-renovate[bot]` (GitHub Enterprise Server).
 
+## Package Registry Credentials
+
+When Renovate is running against repos on `github.com`, it will automatically provision `hostRules` for these GitHub Packages registries using the platform token:
```
Comment (HonkingGoose) [inline lib/modules/platform/github/index.md:90]:
```suggestion
When Renovate runs against repositories on `github.com`, it automatically provisions `hostRules` for these GitHub Packages registries using the platform token:
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/modules/platform/github/index.ts (hunk 1 above)
Comment (rarkins) [reply lib/modules/platform/github/index.ts:222]: image

But I think you're right. I think I was reading here before but got confused: https://github.com/renovatebot/renovate/blob/d53bc5335b4dedef3452a3951f7d1910546945cc/lib/modules/datasource/npm/npmrc.ts#L65-L70

Comment (rarkins) [reply lib/modules/platform/github/index.ts:222]: Fixed

File: docs/usage/getting-started/private-packages.md [hunk 3]
```diff
@@ -476,6 +476,27 @@ module.exports = {
 };
 ```
 
+## Automatic hostRules credentials for platform-hosted registries
+
+### GitHub Packages
+
+For GitHub Packages, Renovate will automatically provision hostRules for both `ghcr.io` (containers) and `*.pkg.github.com` (maven, npm, nuget, rubygems) using the GitHub platform token.
+This means that any private packages hosted in GitHub will be automatically authenticated if they are accessible using the same token by Renovate.
```
Comment (HonkingGoose) [inline docs/usage/getting-started/private-packages.md:484]:
```suggestion
This means that any private packages hosted on GitHub will be automatically authenticated if they are accessible using the same token by Renovate.
```

I think you meant "hosted _on_ GitHub" here?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by HonkingGoose — APPROVED** [review]

File: docs/usage/getting-started/private-packages.md [hunk 4]
```diff
@@ -476,6 +476,28 @@ module.exports = {
...
+  "hostRules": [
+    {
+      "matchHost": "npm.pkg.github.com",
+      "hostType": "npm",
+      "token": "some-personal-access-token"
+    }
+  ]
+}
+```
+
```
[the remark this answers is not in the document]
Comment (rarkins) [reply docs/usage/getting-started/private-packages.md:500]: Is hostType necessary here? And do you think this is really that common?
- renovate won't reach out unless you have a dependency
- it's all GitHub-owned anyway so not any "security risk"

**Review by viceice — APPROVED** [review]

File: docs/usage/getting-started/private-packages.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply docs/usage/getting-started/private-packages.md:500]: I'm going to skip for now as I think it risks confusing users

**Pull request merged**